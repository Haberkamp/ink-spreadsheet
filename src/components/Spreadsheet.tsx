import { reachedEnd } from '../utils/array.utils.js';
import React from 'react';
import useColumnWidth from '../hooks/useColumnWidth.js';
import { Box, Text } from 'ink';
import {
  BORDER_BOTTOM,
  BORDER_STYLE,
  HORIZONTAL_PADDING,
} from '../constants.js';
import { ColumnDef } from '../types.js';

interface Props {
  columns: ColumnDef[];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- needed until we've found a way to provide interfaces to the generic
  data: Record<string, any>[];
}

/**
 * @example
 * const columnHelper = createColumnHelper<Entity>();
 * const columns = [
 *   columnHelper('name'),
 *   columnHelper('email', {
 *     header: 'Email'
 *   }),
 * ];
 *
 * export function MyComponent() {
 *   return (
 *     <Spreadsheet columns={columns} data={[]} />
 *   )
 * }
 */
export default function Spreadsheet({ data, columns }: Props) {
  const columnWidth = useColumnWidth(columns.length);

  return (
    <Box width="100%" flexDirection="column" borderStyle={BORDER_STYLE}>
      <Box borderStyle={BORDER_STYLE} {...BORDER_BOTTOM}>
        {columns.map((column, index, items) => (
          <Box
            key={column.key}
            width={columnWidth}
            paddingX={HORIZONTAL_PADDING}
            borderStyle={BORDER_STYLE}
            borderTop={false}
            borderLeft={false}
            borderRight={!reachedEnd(index, items)}
            borderBottom={false}
          >
            <Text bold>{column.header}</Text>
          </Box>
        ))}
      </Box>

      {data.map((row, index, rows) => (
        <Box
          key={index}
          borderStyle={BORDER_STYLE}
          borderBottom={!reachedEnd(index, rows)}
          borderLeft={false}
          borderRight={false}
          borderTop={false}
        >
          {Object.values(row).map((value, index, items) => (
            <Box
              key={index}
              width={columnWidth}
              paddingX={HORIZONTAL_PADDING}
              borderStyle={BORDER_STYLE}
              borderTop={false}
              borderLeft={false}
              borderRight={!reachedEnd(index, items)}
              borderBottom={false}
            >
              <Text>{value}</Text>
            </Box>
          ))}
        </Box>
      ))}
    </Box>
  );
}
