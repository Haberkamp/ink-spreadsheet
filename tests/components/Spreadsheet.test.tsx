import React from 'react';
import { render } from 'ink-testing-library';
import { describe, it, expect } from 'vitest';
import Spreadsheet from '../../src/components/Spreadsheet.js';

describe('src/component/Spreadsheet.tsx', () => {
  it.each([
    [
      [
        {
          header: 'Full name',
          key: 'name',
        },
        {
          header: 'Email',
          key: 'email',
        },
      ],
      [
        {
          name: 'John Doe',
          email: 'john@doe.com',
        },
        {
          name: 'John Doe',
          email: 'john@doe.com',
        },
      ],
    ],
    // single column
    [
      [
        {
          header: 'Full name',
          key: 'name',
        },
      ],
      [
        {
          name: 'John Doe',
        },
      ],
    ],
    // long cell data
    [
      [
        {
          header: 'Path',
          key: 'path',
        },
        {
          header: 'Size',
          key: 'size',
        },
      ],
      [
        {
          path: '/this/is/a/very/long/path/that/will/be/wrapped/or/will/it/not/?',
          size: '1.2MB',
        },
        {
          path: '/not/so/long/path',
          size: '100GB',
        },
      ],
    ],
  ])('should render the correct output', (columns, data) => {
    const { lastFrame } = render(<Spreadsheet columns={columns} data={data} />);

    expect(lastFrame()).toMatchSnapshot();
  });
});
