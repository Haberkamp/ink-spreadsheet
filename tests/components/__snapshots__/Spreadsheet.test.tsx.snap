// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`src/component/Spreadsheet.tsx > should render the correct output 1`] = `
"+--------------------------------------------------------------------------------------------------+
| [1mFull name[22m                                      | [1mEmail[22m                                           |
|--------------------------------------------------------------------------------------------------|
| John Doe                                       | john@doe.com                                    |
|--------------------------------------------------------------------------------------------------|
| John Doe                                       | john@doe.com                                    |
+--------------------------------------------------------------------------------------------------+"
`;

exports[`src/component/Spreadsheet.tsx > should render the correct output 2`] = `
"+--------------------------------------------------------------------------------------------------+
| [1mFull name[22m                                                                                        |
|--------------------------------------------------------------------------------------------------|
| John Doe                                                                                         |
+--------------------------------------------------------------------------------------------------+"
`;

exports[`src/component/Spreadsheet.tsx > should render the correct output 3`] = `
"+--------------------------------------------------------------------------------------------------+
| [1mPath[22m                                           | [1mSize[22m                                            |
|--------------------------------------------------------------------------------------------------|
| /this/is/a/very/long/path/that/will/be/wrapped | 1.2MB                                           |
| /or/will/it/not/?                              |                                                 |
|--------------------------------------------------------------------------------------------------|
| /not/so/long/path                              | 100GB                                           |
+--------------------------------------------------------------------------------------------------+"
`;

exports[`src/component/Spreadsheet.tsx > should render the correct output when using the column helper 1`] = `
"+--------------------------------------------------------------------------------------------------+
| [1mFull name[22m                                                                                        |
|--------------------------------------------------------------------------------------------------|
| John Doe                                                                                         |
+--------------------------------------------------------------------------------------------------+"
`;
